packages/wallet/src/default-rollup-block-submitter.ts

```diff
@@ -0,0 +1,206 @@
+
+const log = getLogger('rollup-block-submitter')
+
+export class DefaultRollupBlockSubmitter implements RollupBlockSubmitter {
+  public static readonly LAST_CONFIRMED_KEY: Buffer = Buffer.from(
+    'last_confirmed'
+  )
+  public static readonly LAST_SUBMITTED_KEY: Buffer = Buffer.from(
+    'last_submitted'
+  )
+  public static readonly LAST_QUEUED_KEY: Buffer = Buffer.from('last_queued')
+
+  private lastSubmitted: number
+  private lastConfirmed: number
+  private lastQueued: number
+  private blockQueue: RollupBlock[]
+
+  constructor(private db: DB, private readonly rollupContract: Contract) {
+    this.blockQueue = []
+  }
+
+  public async init(): Promise<void> {
+    const [
+      lastSubmittedBuffer,
+      lastConfirmedBuffer,
+      lastQueuedBuffer,
+    ] = await Promise.all([
+      this.db.get(DefaultRollupBlockSubmitter.LAST_SUBMITTED_KEY),
+      this.db.get(DefaultRollupBlockSubmitter.LAST_CONFIRMED_KEY),
+      this.db.get(DefaultRollupBlockSubmitter.LAST_QUEUED_KEY),
+    ])
+
+    this.lastSubmitted = !!lastSubmittedBuffer
+      ? parseInt(lastSubmittedBuffer.toString(), 10)
+      : 0
+    this.lastConfirmed = !!lastConfirmedBuffer
+      ? parseInt(lastConfirmedBuffer.toString(), 10)
+      : 0
+    this.lastQueued = !!lastQueuedBuffer
+      ? parseInt(lastQueuedBuffer.toString(), 10)
+      : 0
+
+    // We're up to date, return
+    if (
+      this.lastSubmitted === this.lastConfirmed &&
+      this.lastConfirmed === this.lastQueued
+    ) {
+      return
+    }
+
+    // We need to populate the queue from storage
+    if (this.lastConfirmed !== this.lastQueued) {
+      let i: number = this.lastConfirmed + 1
+      const promises: Array<Promise<Buffer>> = []
+      for (; i <= this.lastQueued; i++) {
+        promises.push(this.db.get(DefaultRollupBlockSubmitter.getBlockKey(i)))
+      }
+
+      const blocks: Buffer[] = await Promise.all(promises)
+      this.blockQueue = blocks.map((x) =>
+        DefaultRollupBlockSubmitter.deserializeRollupBlockFromStorage(x)
+      )
+    }
+
+    await this.trySubmitNextBlock()
+  }
+
+  public async submitBlock(rollupBlock: RollupBlock): Promise<void> {
+    if (rollupBlock.number <= this.lastQueued) {
+      log.error(
+        `submitBlock(...) called on old block. Last Queued: ${
+          this.lastQueued
+        }, received: ${JSON.stringify(rollupBlock)}`
+      )
+      return
+    }
+
+    this.blockQueue.push(rollupBlock)
+    await this.db.put(
+      DefaultRollupBlockSubmitter.getBlockKey(rollupBlock.number),
+      DefaultRollupBlockSubmitter.serializeRollupBlockForStorage(rollupBlock)
+    )
+
+    this.lastQueued = rollupBlock.number
+    await this.db.put(
+      DefaultRollupBlockSubmitter.LAST_QUEUED_KEY,
+      this.numberToBuffer(this.lastQueued)
+    )
+
+    await this.trySubmitNextBlock()
+  }
+
+  public async handleNewRollupBlock(rollupBlockNumber: number): Promise<void> {
```

Comment:
It's when a rollup block event is emitted, which should contain a block number. There may be multiple rollup blocks in one L1 block, but probably not.

Reply:
It's confusing because this isn't wired up yet, but the RollupAggregator `implements EthereumListener<Event>`, where it will subscribe to RollupChain.sol's `NewRollupBlock(...)` event. Right now that just has all of the transitions, but that will be changed to have the block number (and eventually get rid of the transitions).